components/optimizer/base.py

```diff
@@ -0,0 +1,40 @@
+import abc
+from common.minimodel import Group, Plan
+from components.selector import GroupInfo
+from typing import Dict
+
+Selection = Dict[Group, GroupInfo]
+
+
+class BaseOptimizer(metaclass=abc.ABCMeta):
+    """
+    Base class for all Optimizer components.
+    Each optimizing algorithm needs to implement the following methods:
+
+    schedule: methods that triggers the formation of the plan
+    get_visits: methods that updates the plan 
+
+    """
+
+    @classmethod
+    def __subclasshook__(cls, subclass):
+        return (hasattr(subclass, 'schedule') and
+                callable(subclass.schedule) and
+                hasattr(subclass, 'add') and
+                callable(subclass.add) and
+                hasattr(subclass, 'get_visits') and
+                callable(subclass.get_visits) and
+                hasattr(subclass, '_run') and
+                callable(subclass, '_run'))
+
+    @property
+    def get_visits(self):
+        return self._visits
```

Comment:
I'm not entirely sure that this is the correct use of `@property`.

https://www.programiz.com/python-programming/property

This should probably just be called:
```
@property
def visits(self):
    return self._visits
```
but in this case, it really doesn't serve a purpose, because you're not doing any checking on `self._visits` and you're returning a reference to a list, so if anyone accesses it, they can change it and the changes will be propagated back to the `BaseOptimizer`.

At this point, you may as well just have `self.visits` and not make it "Python private" and allow users to access it directly.

If the intention is to prevent modification, you should have something like:

```
from copy import copy, deepcopy  # Not sure which one you want
...
@property
def visits(self):
    # One of these two:
    return copy(self._visits)
    return deepcopy(self._visits)
```

Reply:
So what I wanted to force was the subclasses to implement a self._visits member. This was removed because visits are now saved in the plan. So no need for this bad code